### PR TITLE
Mark Assignments as Complete

### LINKED ISSUE #90: "Complete" Assignments
Create functionality to mark assignments as complete and render that information to the club admin.

### PR DESCRIPTION
**Pull request by steven-solar**
This PR allows club members to mark assignments as complete, and uncomplete. This allows users to track which assignments they've finished. 

Renders based on which I've completed:
Screenshot 2020-07-29 at 3 26 58 PM

Marking one as completed:
Screenshot 2020-07-29 at 3 27 05 PM

### CODE REVIEW

File: capstone-frontend/package.json [hunk 1]
```diff
@@ -38,5 +37,9 @@
       "last 1 firefox version",
       "last 1 safari version"
     ]
+  },
+  "devDependencies": {
+    "@testing-library/dom": "^7.21.6",
+    "@testing-library/user-event": "^7.2.1"
```
Comment (steven-solar) [inline capstone-frontend/package.json:43]: This fixed a failing npm run build in the Git automated tests.

File: capstone-frontend/src/components/AssignmentCard.jsx [hunk 2]
```diff
@@ -10,7 +10,8 @@ class AssignmentCard extends Component {
   constructor(props) {
     super(props);
     this.state = {
-      comments: []
+      comments: [], 
+      completed: this.props.assignment.completedIDs && this.props.assignment.completedIDs.includes(window.localStorage.getItem('userID')),
```
Comment (antoniolinhart) [inline capstone-frontend/src/components/AssignmentCard.jsx]: nit: This should be broken up into multiple lines

File: capstone-frontend/src/components/AssignmentCard.jsx [hunk 3]
```diff
@@ -45,7 +46,25 @@ class AssignmentCard extends Component {
   }
 
   onComplete = () => {
-    //TODO #90: Create functionality to mark assignments as complete
+    console.log('complete');
```
Comment (antoniolinhart) [inline capstone-frontend/src/components/AssignmentCard.jsx]: nit: Remove console logs

File: capstone-frontend/src/components/AssignmentCard.jsx [hunk 4]
```diff
@@ -45,7 +46,25 @@ class AssignmentCard extends Component {
...
+      id: this.props.assignment.id,
+      add_completedIDs: window.localStorage.getItem('userID')
+    }
+    fetch('/api/assignments', {method:'put', body: JSON.stringify(completeAssignment)})
+        .then(this.setState({completed: true}))
+        .catch(e => alert(e));
+  }
+
+    onUncomplete = () => {
+    console.log('uncomplete');
```
Comment (antoniolinhart) [inline capstone-frontend/src/components/AssignmentCard.jsx]: nit: Same as above

File: capstone-frontend/src/components/UserCard.jsx [hunk 5]
```diff
@@ -12,7 +12,7 @@ export class UserCard extends Component {
     }
     fetch(`/api/user?id=${window.localStorage.getItem('userID')}`)
       .then(response => response.json())
-      .then(user => this.setState({isFriend:user.friendIDs.includes(this.props.user.id)}))
+      .then(user => this.setState({isFriend: user.friendIDs && user.friendIDs.includes(this.props.user.id)}))
```
Comment (antoniolinhart) [inline capstone-frontend/src/components/UserCard.jsx:15]: If user.friendIDs doesn't exist, does this statement mean that the state is automatically false?

**Review by antoniolinhart — APPROVED** [review]
LGTM

**Review by MauroGuerrero20 — APPROVED** [review]
LGTM

File: capstone-frontend/src/components/UserCard.jsx (hunk 5 above)
Comment (steven-solar) [reply capstone-frontend/src/components/UserCard.jsx:15]: Yes it does. It also logically works, as if they have no friends, then this user is definitely not their friend.

File: capstone-backend/src/main/java/com/google/sps/data/Assignment.java [hunk 6]
```diff
@@ -9,16 +9,18 @@
...
   private String whenDue;
+  private List<String> completedIDs;
 
-  public Assignment(String text, String clubID, String whenCreated, String whenDue) {
+  public Assignment(String text, String clubID, String whenCreated, String whenDue, List<String> completedIDs) {
     this.text = text;
     this.clubID = clubID;
     this.whenCreated = whenCreated;
     this.whenDue = whenDue;
+    this.completedIDs = completedIDs;
```
Comment (VivekShankar) [inline capstone-backend/src/main/java/com/google/sps/data/Assignment.java:19]: Nice! Having each assignment hold the user IDs that have completed it make for a really clean solution!

**Review by VivekShankar — APPROVED** [review]
LGTM

File: capstone-backend/src/main/java/com/google/sps/data/Assignment.java (hunk 6 above)
Comment (steven-solar) [reply capstone-backend/src/main/java/com/google/sps/data/Assignment.java:19]: Thanks!

**Pull request merged**